File: Problem7.java
//Time complexity: O(logn)
//Space Complexity: O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

public class Problem7 {
    public int[] searchRange(int[] nums, int target) {
        int leftIndex = leftMost(nums,target);
        int rightIndex = rightMost(nums, target);
        return new int[]{leftIndex, rightIndex};
    }

    public int leftMost(int[] nums, int target){
        int left = 0;
        int right = nums.length - 1;
        while(left<=right){
            int mid = left + (right-left)/2;
            if(nums[mid]==target){
                if(nums[0]==target)
                    return 0;
                else if(nums[mid-1]== target)
                    right = mid-1;
                else
                    return mid;
            }else{
                if(nums[mid]>=target)
                    right = mid-1;
                else
                    left = mid+1;
            }
        }
        return -1;
    }

    public int rightMost(int[] nums, int target){
        int left = 0;
        int right = nums.length - 1;
        while(left<=right){
            int mid = left + (right-left)/2;
            if(nums[mid]==target){
                if(nums[nums.length-1]==target)
                    return nums.length-1;
                else if(nums[mid+1]== target)
                    left = mid+1;
                else
                    return mid;
            }else{
                if(nums[mid]>=target)
                    right = mid-1;
                else
                    left = mid+1;
            }
        }
        return -1;
    }
}

File: Problem7.py
# 34. Find First and Last Position of Element in Sorted Array.

# Time Complexity : O(log n) which is binary search complexity
# Space Complexity : O(1) which is binary search complexity
# Did this code successfully run on Leetcode : Yes, with Runtime: 64 ms and Memory Usage: 14.3 MB
# Any problem you faced while coding this : No
# Your code here along with comments explaining your approach
# Approach
"""
Find target using Binary Search 
There are two choices finding first occurence and last occurence.
On figuring the sorted portion, find the element if it lies in the range otherwise move further
"""

class Solution(object):
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        def first_occurence(nums,target,first_index,last_index):
            if first_index > last_index: 
                return -1
            while(first_index<=last_index):
                middle=(first_index+last_index)//2
                if(target==nums[middle]):
                    if(middle==first_index or nums[middle-1]< target):
                         return middle
                    else:
                        last_index=middle-1
                elif(target > nums[middle]):
                    first_index=middle+1
                else:
                    last_index=middle-1
            return -1
                    
        def last_occurence(nums,target,first_index,last_index):
                if first_index > last_index: 
                    return -1
                while(first_index<=last_index):
                    middle=(first_index+last_index)//2
                    if(target==nums[middle]):
                        if(middle==last_index or target<nums[middle+1]):
                            return middle
                        else:
                            first_index=middle+1
                    elif(target > nums[middle]):
                        first_index=middle+1
                    else:
                        last_index=middle-1
                return -1
        first_index=0
        last_index=len(nums)-1
        first_o=first_occurence(nums,target,first_index,last_index)
        last_o=last_occurence(nums,target,first_index,last_index)   
        return [first_o,last_o]
                    

            